File: scripts/agents/hooks/pre-tool-guard.mjs
#!/usr/bin/env node
/**
 * Dependency-free cross-provider PreToolUse guard.
 *
 * stdin: provider hook payload JSON
 * allow: no output, exit 0
 * block: stderr reason, exit 2
 */
import { createHash } from 'node:crypto';
import { appendFileSync, mkdirSync, readFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..', '..', '..');
export const AUDIT_LOG =
  process.env.HUMANYMOUS_HOOK_LOG || join(ROOT, '.agent-runs', 'hooks', 'pre-tool-guard.jsonl');

const BLOCK_PATTERNS = [
  [/\bcurl\s+.*\|\s*(?:ba)?sh\b/i, 'blocked: curl|sh pipe'],
  [
    /\bnmap\s+(?:-[a-z0-9]+\s+){0,6}(?!127\.|localhost|10\.|192\.168\.|172\.(?:1[6-9]|2[0-9]|3[0-1])\.)\d/i,
    'blocked: nmap against non-local target',
  ],
  [
    /\b(?:sqlmap|hydra|nikto)\b.*https?:\/\/(?!127\.|localhost|\[::1\])/i,
    'blocked: offensive tool against non-local URL',
  ],
];

const PROTECTED_BRANCH = /(?:^|[^\w/-])(?:main|master)(?![\w/-])/i;
const FORCE_FLAG = /(?:^|\s)(?:--force(?:-with-lease)?|-f)(?=$|\s)/i;
const ROOT_TARGET =
  /(?:^|\s)(?:--?\w+\s+)*["']?(?:\/|~|\$home|\$env:(?:userprofile|homedrive)|[a-z]:[\\/]+)["']?(?=$|\s)/i;

function sha256(value) {
  return createHash('sha256').update(value).digest('hex');
}

function audit(raw, payload, command, reason) {
  const entry = {
    version: 1,
    timestamp: new Date().toISOString(),
    pid: process.pid,
    ppid: process.ppid,
    hookEvent: typeof payload?.hook_event_name === 'string' ? payload.hook_event_name : null,
    toolName: typeof payload?.tool_name === 'string' ? payload.tool_name : null,
    inputBytes: Buffer.byteLength(raw),
    inputSha256: sha256(raw),
    commandBytes: Buffer.byteLength(command),
    commandSha256: command ? sha256(command) : null,
    decision: reason ? 'block' : 'allow',
    rule: reason,
    exitCode: reason ? 2 : 0,
    stdoutBytes: 0,
    stderrBytes: reason ? Buffer.byteLength(`${reason}\n`) : 0,
  };

  try {
    mkdirSync(dirname(AUDIT_LOG), { recursive: true });
    appendFileSync(AUDIT_LOG, `${JSON.stringify(entry)}\n`, 'utf8');
  } catch {
    // Observability must never change the safety decision or hook output.
  }
}

export function extractCommand(payload) {
  if (!payload || typeof payload !== 'object' || Array.isArray(payload)) return '';

  for (const key of ['command', 'cmd', 'shell_command']) {
    if (typeof payload[key] === 'string') return payload[key];
  }

  const toolInput = payload.tool_input ?? payload.input ?? {};
  if (toolInput && typeof toolInput === 'object' && !Array.isArray(toolInput)) {
    for (const key of ['command', 'cmd', 'shell_command']) {
      if (typeof toolInput[key] === 'string') return toolInput[key];
    }
  }

  if (('tool_name' in payload || 'tool' in payload) && typeof payload.tool_input === 'string') {
    return payload.tool_input;
  }
  return '';
}

export function evaluateCommand(command) {
  const normalized = String(command).trim().split(/\s+/).join(' ');
  const lower = normalized.toLowerCase();

  if (/\bgit\s+push\b/.test(lower) && FORCE_FLAG.test(normalized) && PROTECTED_BRANCH.test(normalized)) {
    return 'blocked: force-push to main/master';
  }

  if (/\brm\b/.test(lower)) {
    const recursive = /(?:^|\s)(?:-[a-z]*r[a-z]*|--recursive)(?=$|\s)/i.test(lower);
    const force = /(?:^|\s)(?:-[a-z]*f[a-z]*|--force)(?=$|\s)/i.test(lower);
    if (recursive && force && ROOT_TARGET.test(normalized)) {
      return 'blocked: recursive delete of filesystem/home root';
    }
  }

  if (/\bremove-item\b/.test(lower)) {
    const recursive = /(?:^|\s)-(?:recurse|r)(?=$|\s)/i.test(lower);
    if (recursive && ROOT_TARGET.test(normalized)) {
      return 'blocked: PowerShell recursive delete of filesystem/home root';
    }
  }

  if (/(?:^|[^\w-])(?:rd|rmdir)\b/.test(lower)) {
    const recursive = /(?:^|\s)\/s(?=$|\s)/i.test(lower);
    if (recursive && ROOT_TARGET.test(normalized)) {
      return 'blocked: cmd recursive delete of filesystem root';
    }
  }

  for (const [pattern, reason] of BLOCK_PATTERNS) {
    if (pattern.test(normalized)) return reason;
  }
  return null;
}

export function main(raw = readFileSync(0, 'utf8')) {
  let payload;
  if (raw.trim()) {
    try {
      payload = JSON.parse(raw);
    } catch {
      payload = { command: raw };
    }
  } else {
    payload = {};
  }

  const command = extractCommand(payload);
  const reason = command ? evaluateCommand(command) : null;
  audit(raw, payload, command, reason);
  if (!reason) return 0;

  process.stderr.write(`${reason}\n`);
  return 2;
}

if (process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1]) {
  process.exitCode = main();
}

File: scripts/agents/hooks/pre-tool-guard.py
#!/usr/bin/env python3
"""Pre-tool guard for coding agents (Claude Code / Grok Build compatible).

Reads JSON on stdin (tool payload). Exit codes:
  0 — allow
  2 — block (message on stderr; Claude PreToolUse convention)

Blocks high-risk patterns: recursive delete of roots, third-party offensive
scanning framed as external targets, force-push to main/master.
Defense-only redteam against localhost/compose remains allowed.
"""
from __future__ import annotations

import json
import re
import sys

# Commands that should never run without explicit human override outside this guard.
BLOCK_PATTERNS = [
    (re.compile(r"\brm\s+(-[a-zA-Z]*r[a-zA-Z]*f|-[a-zA-Z]*f[a-zA-Z]*r)\s+[/\\]?\s*$"), "blocked: recursive delete of filesystem root"),
    (re.compile(r"\brm\s+(-[a-zA-Z]*r[a-zA-Z]*f|-[a-zA-Z]*f[a-zA-Z]*r)\s+/($|\s)"), "blocked: rm -rf /"),
    (re.compile(r"\bgit\s+push\s+.*--force.*\b(main|master)\b"), "blocked: force-push to main/master"),
    (re.compile(r"\bcurl\s+.*\|\s*(ba)?sh\b"), "blocked: curl|sh pipe"),
    # Obvious third-party offensive framing (not local compose/localhost).
    (re.compile(r"\bnmap\s+(-[a-zA-Z0-9]+\s+){0,6}(?!127\.|localhost|10\.|192\.168\.|172\.(1[6-9]|2[0-9]|3[0-1])\.)\d"), "blocked: nmap against non-local target"),
    (re.compile(r"\b(sqlmap|hydra|nikto)\b.*https?://(?!127\.|localhost|\[::1\])", re.I), "blocked: offensive tool against non-local URL"),
]


def extract_command(payload: dict) -> str:
    # Claude / Grok-style shapes
    for key in ("command", "cmd", "shell_command"):
        if key in payload and isinstance(payload[key], str):
            return payload[key]
    tool_input = payload.get("tool_input") or payload.get("input") or {}
    if isinstance(tool_input, dict):
        for key in ("command", "cmd", "shell_command"):
            if key in tool_input and isinstance(tool_input[key], str):
                return tool_input[key]
    # Nested tool_name + tool_input
    if "tool_name" in payload or "tool" in payload:
        ti = payload.get("tool_input") or {}
        if isinstance(ti, str):
            return ti
    return json.dumps(payload)


def main() -> int:
    raw = sys.stdin.read()
    if not raw.strip():
        return 0
    try:
        payload = json.loads(raw)
    except json.JSONDecodeError:
        # Non-JSON stdin: treat whole line as command string
        payload = {"command": raw}

    cmd = extract_command(payload if isinstance(payload, dict) else {})
    if not cmd:
        return 0

    for pattern, reason in BLOCK_PATTERNS:
        if pattern.search(cmd):
            sys.stderr.write(reason + "\n")
            # Claude PreToolUse: exit 2 blocks
            return 2
    return 0


if __name__ == "__main__":
    sys.exit(main())
